### PR TITLE
Bom 26 chat be

### PR DESCRIPTION
**Pull request by O31**
Changed websocket message a little bit, now it send a type and a payload, which can be of any of the previous types that we had before. Added a SPA router(tried to use the one from the miniwork, didnt function so well) - so we dont need to reconnect the websocket every time the user is redirected. Added user validating in a form that if they dont have username or gameID in the sessionstorage, then they are redirected. For the server side, its easier to complete it when we have a game loop and we validate the players then

### CODE REVIEW

File: frontend/src/models/constants.ts [hunk 1]
```diff
@@ -2,8 +2,10 @@ export enum WsMessageTypes {
     //TODO: map all ws events
     Connect = "connect",
     EnterLobby = "enterLobby",
+    lobbyJoinSuccess = "lobbyJoinSuccess",
```
Comment (khaletska) [inline frontend/src/models/constants.ts:5]: same case please

File: frontend/src/models/wsMessage.ts [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    sender:  string,
+};
+
+type ConnectServerMessage = {
+    clientID: string
+}
+
+//user info who joined or left lobby needed as well?
+export type EnterLobbyServerMessage = {
+    playerCount: number,
```
Comment (khaletska) [inline frontend/src/models/wsMessage.ts:30]: please check map class in BOM-whatever_map_rewriting and think if we can send map with everything in most cases?

**Comment by karmeb** [comment]
Found one loophole: 
If user goes directly to http://localhost:1234/waiting-room, ws connection is established before user enters username. So it is possible to trigger counter and enter the game after counter without entering any username. 

Related to that, when the redirections happen (when user goes to /waiting-room or /game direclty), then there is an error in console after rediretion to "/":
image


And one observation: 
After the changes, it does not look like the username is saved to Player object at any point in the BE. Do we need it there? if not we can just remove the property from Player object and also remove the code from frontend that sends username with empty message type.

**Review by karmeb — APPROVED** [review]
Yes, works now, no loopholes detected anymore :)

**Review by khaletska — APPROVED** [review]

**Comment by khaletska** [comment]
I have one note for future: we don't have any checks for the same usernames in one game. But it might be ok if we show the username in the same color as a player

File: frontend/src/models/constants.ts (hunk 1 above)
Comment (O31) [reply frontend/src/models/constants.ts:5]: i think this one i just forgot to delete

**Pull request merged**